flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/JobReconciler.java

```diff
@@ -86,23 +86,21 @@ public void reconcile(
                     upgradeFlinkJob(flinkApp, effectiveConfig);
                 }
                 if (desiredJobState.equals(JobState.SUSPENDED)) {
-                    if (upgradeMode == UpgradeMode.STATELESS) {
-                        cancelJob(flinkApp, effectiveConfig);
-                    } else {
-                        suspendJob(flinkApp, effectiveConfig);
-                    }
+                    printCancelLogs(upgradeMode, flinkApp.getMetadata().getName());
+                    cancelJob(flinkApp, upgradeMode, effectiveConfig);
                 }
             }
-            if (currentJobState == JobState.SUSPENDED) {
-                if (desiredJobState == JobState.RUNNING) {
-                    if (upgradeMode == UpgradeMode.STATELESS) {
-                        deployFlinkJob(flinkApp, effectiveConfig, Optional.empty());
-                    } else if (upgradeMode == UpgradeMode.SAVEPOINT) {
-                        restoreFromLastSavepoint(flinkApp, effectiveConfig);
-                    } else {
-                        throw new InvalidDeploymentException(
-                                "Only savepoint and stateless strategies are supported at the moment.");
-                    }
+            if (currentJobState == JobState.SUSPENDED && desiredJobState == JobState.RUNNING) {
+                if (upgradeMode == UpgradeMode.STATELESS) {
+                    deployFlinkJob(flinkApp, effectiveConfig, Optional.empty());
+                } else if (upgradeMode == UpgradeMode.SAVEPOINT) {
+                    restoreFromLastSavepoint(flinkApp, effectiveConfig);
+                } else if (upgradeMode == UpgradeMode.LAST_STATE) {
+                    final String savepointLocation =
+                            flinkApp.getStatus().getJobStatus().getSavepointLocation();
+                    // Upgrade mode changes from savepoint -> last-state
+                    deployFlinkJob(
+                            flinkApp, effectiveConfig, Optional.ofNullable(savepointLocation));
```

Comment:
IMO, we may need to clean the `savepointLocation` after we cancel the job with `LAST_STATE`, Otherwise, if we have done once `SUSPEND` at `SAVEPOINT` mode, this path persistent in the `JobStatus` It will always be the first choice  in `LAST_STATE` 
 upgrade mode.

About the original case posted by @gyfora 

>  if the user suspends a job with savepoint upgrademode and then changes the upgrade mode to last-state when resubmitting it to running state.
In this case with the current logic the job would simply start from an empty state which is incorrect

I have a question here, why it will start from an empty state? It will start from the latest checkpoint from HA data, It only start from an empty state when there have not generated one checkpoint.

Reply:
If the user suspends a job with savepoint upgrademode, the HA ConfigMaps will be cleaned up as well.

Reply:
> IMO, we may need to clean the `savepointLocation` after we cancel the job with `LAST_STATE`, Otherwise, if we have done once `SUSPEND` at `SAVEPOINT` mode, this path persistent in the `JobStatus` It will always be the first choice in `LAST_STATE` upgrade mode.

Having the savepoint path in the `JobStatus` is harmless. If we specify both savepoint and HA config options, Flink will tries to pick the latest one. We will introduce manual or periodical savepoint in the future, I think having the savepoint path in the `JobStatus` will make more senses then.

Reply:
> If we specify both savepoint and HA config options, Flink will try to pick the latest one.

I have no detailed background knowledge for this, I will take a look for it then. Make sense to me, thanks for explanation! +1 for this 👍